pkg/query-service/app/logs/v4/resource_query_builder.go

```diff
@@ -0,0 +1,204 @@
+
+	chFmtVal := utils.ClickHouseFormattedValue(value)
+
+	switch op {
+	case v3.FilterOperatorExists:
+		return fmt.Sprintf("simpleJSONHas(lower(labels), '%s')", key)
+	case v3.FilterOperatorNotExists:
+		return fmt.Sprintf("not simpleJSONHas(lower(labels), '%s')", key)
+	case v3.FilterOperatorRegex, v3.FilterOperatorNotRegex:
+		return fmt.Sprintf(logsOp, searchKey, chFmtVal)
+	case v3.FilterOperatorContains, v3.FilterOperatorNotContains:
+		// this is required as clickhouseFormattedValue add's quotes to the string
+		lowerEscapedStringValue := utils.QuoteEscapedString(strings.ToLower(fmt.Sprintf("%s", value)))
+		return fmt.Sprintf("%s %s '%%%s%%'", searchKey, logsOp, lowerEscapedStringValue)
+	default:
+		chFmtValLower := strings.ToLower(chFmtVal)
+		return fmt.Sprintf("%s %s %s", searchKey, logsOp, chFmtValLower)
+	}
+}
+
+// buildIndexFilterForInOperator builds a clickhouse filter string for in operator
+// example:= x in a,b,c = (labels like '%x%a%' or labels like '%"x":"b"%' or labels like '%"x"="c"%')
+// example:= x nin a,b,c = (labels nlike '%x%a%' AND labels nlike '%"x"="b"' AND labels nlike '%"x"="c"%')
+func buildIndexFilterForInOperator(key string, op v3.FilterOperator, value interface{}) string {
+	conditions := []string{}
+	separator := " OR "
+	sqlOp := "like"
+	if op == v3.FilterOperatorNotIn {
+		separator = " AND "
+		sqlOp = "not like"
+	}
+
+	// values is a slice of strings, we need to convert value to this type
+	// value can be string or []interface{}
+	values := []string{}
+	switch value.(type) {
+	case string:
+		values = append(values, value.(string))
+	case []interface{}:
+		for _, v := range (value).([]interface{}) {
+			// also resources attributes are always string values
+			strV, ok := v.(string)
+			if !ok {
+				continue
+			}
+			values = append(values, strV)
+		}
+	}
+
+	// if there are no values to filter on, return an empty string
+	if len(values) > 0 {
+		for _, v := range values {
+			conditions = append(conditions, fmt.Sprintf("lower(labels) %s '%%\"%s\":\"%s\"%%'", sqlOp, key, strings.ToLower(v)))
+		}
+		return "(" + strings.Join(conditions, separator) + ")"
+	}
+	return ""
+}
+
+// buildResourceIndexFilter builds a clickhouse filter string for resource labels
+// example:= x like '%john%' = lower(labels) like '%x%john%'
+func buildResourceIndexFilter(key string, op v3.FilterOperator, value interface{}) string {
+	// not using clickhouseFormattedValue as we don't wan't the quotes
+	formattedValueEscapedLower := utils.QuoteEscapedString(strings.ToLower(fmt.Sprintf("%s", value)))
+
+	// add index filters
+	switch op {
+	case v3.FilterOperatorContains, v3.FilterOperatorEqual, v3.FilterOperatorLike:
+		return fmt.Sprintf("lower(labels) like '%%%s%%%s%%'", key, formattedValueEscapedLower)
+	case v3.FilterOperatorNotContains, v3.FilterOperatorNotEqual, v3.FilterOperatorNotLike:
+		return fmt.Sprintf("lower(labels) not like '%%%s%%%s%%'", key, formattedValueEscapedLower)
```

Comment:
How do these work when value has `%` or `_` in them?

Reply:
It should keep them as it is, 

i.e `x contains 'app%_test`

Should result in
`lower(labels) ilike  '%x%app%_test%'`

Adding a test case for this

Reply:
have added a test case for this `test contains with % and _`

Reply:
`contains` is regular string matching. It should not keep them as is. If I am searching for `service.name` contains `my_service` but the `like` or `not like` will not treat `_`, `%` as literals.

Reply:
got your point, even though the result will be there, better to keep expectations correct.